File: programmers/모의고사.py
"""
코딩테스트 연습 > 완전탐색
"""


def solution(answers):
    num = len(answers)

    ans_1 = [1, 2, 3, 4, 5] * (num // 5 + 1)
    ans_2 = [2, 1, 2, 3, 2, 4, 2, 5] * (num // 8 + 1)
    ans_3 = [3, 3, 1, 1, 2, 2, 4, 4, 5, 5] * (num // 10 + 1)

    scores = [[0, 0], [1, 0], [2, 0]]

    for i in range(num):
        answer = answers[i]
        if answer == ans_1[i]:
            scores[0][1] += 1
        if answer == ans_2[i]:
            scores[1][1] += 1
        if answer == ans_3[i]:
            scores[2][1] += 1

    top_score = max(scores, key = lambda x: x[1])[1]

    result = list(map(lambda x: x[0]+1, filter(lambda x: x[1] == top_score, scores)))

    return result

File: programmers/모의고사.kt
class Solution {
    fun solution(answers: IntArray): IntArray {
        var pA = "12345".repeat(answers.size / 5 + 1).toList()
        var pB = "21232425".repeat(answers.size / 8 + 1).toList()
        var pC = "3311224455".repeat(answers.size / 10 + 1).toList()

        var scores = mutableListOf(0, 0, 0)

        answers.mapIndexed { i, v ->
            if (pA[i].toInt() - '0'.toInt() == v) {
                scores[0] += 1
            }

            if (pB[i].toInt() - '0'.toInt() == v) {
                scores[1] += 1
            }

            if (pC[i].toInt() - '0'.toInt() == v) {
                scores[2] += 1
            }
        }

        var maxV = scores.maxOf{it}
        var result = mutableListOf<Int>()

        scores.mapIndexed { i, v ->
            if (v == maxV) {
                result.add(i+1)
            }
        }

        return result.toIntArray()

    }
}
